iron-fish/ironfish
# Rahul/ifl 1757 address manager improvements

## Summary

1. We save peers in the hosts.json file when their connection gets updated to CONNECTED and is a websocket + outbound connection
2. We don't overwrite the peers on disk any more, we append to the list instead of truncating it
3. We connect to those peers on startup first (and that's it)
4. Added the field `lastAddedTimestamp` that is used to sort the peers we want to keep in the address manager (by most recent)
5. We are no longer saving candidates on a timer. We just write them down as we find good candidates. We also write it down once when the `stop` command is called. On testing, when we connect to 50 peers (which took my machine 4 hours at home), we usually connect to around 10 candidates, the number of writes to disk is low

## Metric Improvements

Test: 
1. Get a fully synced node
2. Remove the hosts.json file 
3. Measure time to connect to 20 peers
4. Stop node 
6. Measure time again to connect to 20 peers
[I repeated this test manually 4 times]

Setup: Macbook Pro M2 with a 900 mb/s internet connection at my home

Number of peers in the `hosts.json` file after connecting to 20 peers: 
1. current change: ~`5`
2. old address manager: ~`20`

Time to get to 20 peers without hosts.json: around `16 minutes`

Time to get to 20 peers w/o bootstrap nodes in the hosts.json file and using the `--bootstrap = ''` flag: 
1. current change: `Low end: 3 minutes. High end: 12 minutes`
2. old address manager: `Low end: 14 minutes. High end: +30 minutes`. Note: there were some cases it would connect to a bad peer off the bat and just stay stuck at 1 peer. The likelihood of this happening with the current change is drastically reduced. 

This is not a perfect test because luck is a significant part of the equation at the moment. You might just happen to connect with a really good or bad peer off the bat and that might derail the test. However, the current change does consistently connect to good peers and also reduces the reliance on the bootstrap nodes. This test might have also been impacted by my router's settings/ internet provider.

## Testing Plan
Updating unit tests

Manual: 
1. Run the node once with this code and accumulate about 4 - 8 peers in the `hosts.json` file
7. Run the node again but with `--bootstrap ''` flag and you will notice a speed up on connecting to 3-8 peers off the bat. 
I also did the above in a wework when I had a solid set of peers. It significantly sped up connecting to peers.

## Documentation

Does this change require any updates to the Iron Fish Docs (ex. [the RPC API
Reference](https://ironfish.network/docs/onboarding/rpc/chain))? If yes, link a
related documentation pull request for the website.

```
[ ] Yes
```

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and
what additional work is required, if any.

```
[ ] Yes
```
